src/main/java/com/gradle/enterprise/bamboo/config/PluginEnabledListener.java

```diff
@@ -0,0 +1,47 @@
+package com.gradle.enterprise.bamboo.config;
+
+import com.atlassian.bamboo.bandana.PlanAwareBandanaContext;
+import com.atlassian.bandana.BandanaManager;
+import com.atlassian.event.api.EventListener;
+import com.atlassian.plugin.event.events.PluginEnabledEvent;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PluginEnabledListener {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PluginEnabledListener.class);
+    private static final String PLUGIN_KEY = "com.gradle.enterprise.gradle-enterprise-bamboo-plugin";
+    private final BandanaManager bandanaManager;
+    private final JsonConfigurationConverter jsonConfigurationConverter;
+
+    @Autowired
+    public PluginEnabledListener(BandanaManager bandanaManager, JsonConfigurationConverter jsonConfigurationConverter) {
+        this.bandanaManager = bandanaManager;
+        this.jsonConfigurationConverter = jsonConfigurationConverter;
+    }
+
+    @EventListener
+    public void onPluginEnabled(PluginEnabledEvent event) {
+        if (PLUGIN_KEY.equals(event.getPlugin().getKey())) {
+            migrateLegacyConfigToV1();
+        }
+    }
+
+    private void migrateLegacyConfigToV1() {
+        Object value = bandanaManager.getValue(PlanAwareBandanaContext.GLOBAL_CONTEXT, PersistentConfigurationManager.LEGACY_CONFIG_V0_KEY);
+        if (value != null && PersistentConfiguration.class.getName().equals(value.getClass().getName())) {
+            try {
+                LOGGER.info("Migrating {} config to json", PLUGIN_KEY);
+                String json = jsonConfigurationConverter.toJson(value);
```

Comment:
If I understand it correctly, this event `PluginEnabledEvent` will be sent after the plugin is reloaded, so we are already in the new classloader. We can't cast the retrieved object to `PersistentConfiguration`, but this trick with Jackson works. Why then simply not "migrate on get" using this trick? Then we don't need another key, we don't need another listener. Just get an object, if it's not a string, run it through Jackson, store and retrieve again.

Reply:
Yes, I began with this approach but I found it harder to follow and mixing concerns, specially because the issue is kind of hard to trigger. Hence I've created this little indirection to migrate the config. The config manager should be based on the pre-condition that everything is in the expected format.